src/JSInterop/Microsoft.JSInterop/src/DotNetDispatcher.cs

```diff
@@ -133,54 +146,96 @@ private static object InvokeSynchronously(string assemblyName, string methodIden
+                var index = 0;
+                foreach (var item in jsonDocument.RootElement.EnumerateArray())
                 {
-                    suppliedArgs[i] = serializerStrategy.DeserializeObject(
-                        suppliedArgs[i], parameterTypes[i]);
+                    var parameterType = parameterTypes[index];
+
+                    if (parameterType == typeof(JSAsyncCallResult))
+                    {
+                        // We will pass the JsonDocument instance to JAsyncCallResult and make JSRuntimeBase
+                        // responsible for disposing it.
+                        shouldDisposeJsonDocument = false;
+                        // For JS async call results, we have to defer the deserialization until
+                        // later when we know what type it's meant to be deserialized as
+                        suppliedArgs[index] = new JSAsyncCallResult(jsonDocument, item);
+                    }
+                    else if (IsIncorrectDotNetObjectRefUse(item, parameterType))
+                    {
+                        throw new InvalidOperationException($"In call to '{methodIdentifier}', parameter of type {parameterType.Name} at index {(index + 1)} must be wrapped in a {nameof(DotNetObjectRef<object>)} instance.");
+                    }
+                    else
+                    {
+                        suppliedArgs[index] = JsonSerializer.Parse(item.GetRawText(), parameterType, _jsonSerializerOptions);
+                    }
+
+                    index++;
                 }
-            }
 
-            try
-            {
-                return methodInfo.Invoke(targetInstance, suppliedArgs);
-            }
-            catch (TargetInvocationException tie) // Avoid using exception filters for AOT runtime support
-            {
-                if (tie.InnerException != null)
+                if (shouldDisposeJsonDocument)
                 {
-                    ExceptionDispatchInfo.Capture(tie.InnerException).Throw();
-                    throw null; // unreached
+                    jsonDocument.Dispose();
                 }
 
+                return suppliedArgs;
+            }
+            catch
+            {
+                // Always dispose the JsonDocument in case of an error.
+                jsonDocument.Dispose();
                 throw;
             }
+
+
+            static bool IsIncorrectDotNetObjectRefUse(JsonElement item, Type parameterType)
+            {
+                // Check for incorrect use of DotNetObjectRef<T> at the top level. We know it's
+                // an incorrect use if there's a object that looks like { '__dotNetObject': <some number> },
+                // but we aren't assigning to DotNetObjectRef{T}.
+                return item.Type == JsonValueType.Object &&
+                     !typeof(IDotNetObjectRef).IsAssignableFrom(parameterType) &&
+                     item.TryGetProperty(DotNetObjectRefKey, out _);
```

Comment:
it would be better to make L237 the second condition and L236 the last one. `IsAssignableFrom` is expensive.